File: 3217-delete-nodes-from-linked-list-present-in-array/3217-delete-nodes-from-linked-list-present-in-array.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* modifiedList(int* nums, int numsSize, struct ListNode* head) {
    int *table = (int *)calloc(sizeof(int), 100001);
    for (int i = 0; i < numsSize; table[nums[i++]]++);

    struct ListNode *pre = NULL, *curr = head;

    while (curr != NULL){
        if ( table[(curr -> val)] ){
            if (curr == head){
                pre = head;
                head = head -> next;
                curr = head;
            }
            else{
                pre -> next = curr -> next;
                free(curr);
                curr = pre -> next;
            }
        }

        else {
            pre = curr;
            curr = curr -> next;           
        }
    }

    free(table);

    return head;
}

File: 3217-delete-nodes-from-linked-list-present-in-array/3217-delete-nodes-from-linked-list-present-in-array.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode modifiedList(int[] nums, ListNode head) {
        HashSet<Integer>set=new HashSet<>();
        for (int i = 0; i < nums.length; set.add(nums[i++]));

        ListNode pre = null;
        ListNode curr = head;

        while (curr != null){
            if (set.contains(curr.val)){
                if (curr == head){
                    head = head.next;
                    curr = head;
                }
                else{
                    pre.next = curr.next;
                    curr = pre.next;
                }
            }
            else{
                pre = curr;
                curr = curr.next;
            }
        }
        return head;
    }
}